File: vbKhelu.cpp
#include<bits/stdc++.h>
using namespace std;

typedef long long ll;
// Output the number of numbers in a range L - R whose sum of digits is a perfect square and which has atleast one recurring digit
vector<ll> s;
ll N, T, L, R;
ll dp[14][975][2][1024][2];

#define gc getchar_unlocked

void scanint(ll &x){
    register ll c = gc();
    x = 0;
    for(;(c<48 || c>57);c = gc());
    for(;c>47 && c<58;c = gc()) {x = (x<<1) + (x<<3) + c - 48;}
}

bool isPerfectSquare(ll x){
    ll root = sqrt(x);
    return x != 0 && root* root == x;
}

bool PS[976];

ll f(ll i, ll sumSq, ll tight, ll B, ll rec){
    // i is the current dig, sumSq is the current squareSum
    // tight: whether ith dig in string is equal to ith of s
    // rec: whether recurring digit has occured
    ll ans = dp[i][sumSq][tight][B][rec];
    if (ans != -1)
        return ans;
    ans = 0;
    if (i == N){
        if (rec && PS[sumSq])
            ans = 1;
        else ans = 0;
    }
    else{
        if (tight == 1){
            for (ll dig = 0; dig <= s[i]; dig++){
                if (i == 0 && dig == 0)
                    continue;

                if (dig == s[i]){
                    if ((B >> dig) & 1)
                        ans += f(i + 1, sumSq + dig*dig, 1, B, 1);
                    else
                        ans += f(i + 1, sumSq + dig*dig, 1, B | (1 << dig), rec);
                }
                else{
                    if ((B >> dig) & 1)
                        ans += f(i + 1, sumSq + dig*dig, 0, B, 1);
                    else
                        ans += f(i + 1, sumSq + dig*dig, 0, B | (1 << dig), rec);
                }
            }
        }
        else{
            for (ll dig = 0; dig < 10; dig++){
                // check if this digit has occured
                if ((B >> dig) & 1)
                    ans += f(i + 1, sumSq + dig*dig, 0, B, 1);
                else
                    ans += f(i + 1, sumSq + dig*dig, 0, B | (1 << dig), rec);
            }
        }
    }
    dp[i][sumSq][tight][B][rec] = ans;
    return ans;
}

ll solve(ll num){
    // Solves problem for range 1 - num
    s.clear();
    while (num){
        s.push_back(num % 10);
        num /= 10;
    }
    memset(dp, -1, sizeof(dp));
    reverse(s.begin(), s.end());
    N = s.size();
    return f(0, 0, 1, 0, 0);
}

int main(){
    for (ll i = 0; i < 976; i++)
        PS[i] = isPerfectSquare(i);
    scanint(T);

    while (T--){
        scanint(L);
        scanint(R);
        printf("%lld\n",solve(R) - solve(L -1));
    }


    return 0;
}

File: code/vbKhelu.py
def perSq(x):
    if x == 0:
        return 0
    return int(x ** 0.5) ** 2 == x

s, N = "", 0
dp = []


def get(i, sumSq, tight, B, rec):
    return i * 14 + sumSq * 975 + tight * 2 + B * 1024 + rec * 2


def f(i, sumSq, tight, B, rec):
    ind = get(i, sumSq, tight, B, rec)
    ans = dp[ind]
    if ans != -1:
        return ans
    ans = 0
    if i == N:
        if PS[sumSq] and rec == 1:
            ans = 1
        else:
            ans = 0
    else:
        if tight == 1:
            # last dig
            for dig in xrange(0, s[i] + 1):
                if i == 0 and dig == 0:
                    continue
                if dig == s[i]:
                    # check if dig is already set
                    if (B >> dig) & 1:
                        ans += f(i + 1, dig * dig + sumSq, 1, B, 1)
                    else:
                        # set the digit
                        ans += f(i + 1, dig * dig + sumSq,
                                 1, B | (1 << dig), rec)
                else:
                    if (B >> dig) & 1:
                        ans += f(i + 1, dig * dig + sumSq, 0, B, 1)
                    else:
                        # set the digit
                        ans += f(i + 1, dig * dig + sumSq,
                                 0, B | (1 << dig), rec)
        else:
            for dig in xrange(0, 10):
                if (B >> dig) & 1:
                    ans += f(i + 1, dig * dig + sumSq, 0, B, 1)
                else:
                    # set the digit
                    ans += f(i + 1, dig * dig + sumSq,
                             0, B | (1 << dig), rec)

    dp[ind] = ans
    return ans


def solve(L):
    global s, N, dp
    s = map(int, str(L))
    N = len(s)
    dp = [-1] * (14 * 975 * 2 * 1024 * 2)
    return f(0, 0, 1, 0, 0)

if __name__ == '__main__':
    PS = [1 if perSq(i) else 0 for i in xrange(976)]

    for _ in xrange(input()):
        L, R = map(int, raw_input().split())
        print solve(R) - solve(L - 1)
